src/analysis.rs

```diff
@@ -0,0 +1,547 @@
+use syntax::hir::{
+    Hir, HirKind, Literal, ClassBytes, ClassBytesRange,
+    Class, Visitor, RepetitionRange, RepetitionKind
+};
+use syntax::hir;
+use utf8_ranges::Utf8Sequences;
+
+/// True iff the given expression is one-pass
+pub fn is_onepass(expr: &Hir) -> bool {
+    hir::visit(expr, IsOnePassVisitor::new()).unwrap()
+}
+
+struct IsOnePassVisitor(bool);
+
+impl Visitor for IsOnePassVisitor {
+    type Output = bool;
+    type Err = ();
+
+    fn finish(self) -> Result<bool, ()> {
+        Ok(self.0)
+    }
+
+    fn visit_pre(&mut self, hir: &Hir) -> Result<(), ()> {
+        if !self.0 {
+            return Ok(())
+        }
+
+        match hir.kind() {
+            &HirKind::Concat(ref es) => self.check_concat(&es),
+            &HirKind::Alternation(ref es) => self.check_alternation(&es),
+            &HirKind::Repetition(ref rep) => {
+                if fset_of(&*rep.hir).is_empty() {
+                    self.0 = false;
+                }
+            }
+            &HirKind::Class(ref cls) => self.check_cls(cls),
+            _ => ()
+        }
+
+        Ok(())
+    }
+}
+
+impl IsOnePassVisitor {
+    fn new() -> Self {
+        IsOnePassVisitor(true)
+    }
+
+    fn check_concat(&mut self, es: &[Hir]) {
+        let mut empty_run = vec![];
+
+        for e in NestedConcat::new(es) {
+            // TODO(ethan):yakshaving factor the determination of when
+            //     a regex accepts_empty out into a separate function,
+            //     so that we don't compute the whole first set when we
+            //     don't need to.
+            let fset = fset_of(e);
+            let is_rep = match e.kind() {
+                &HirKind::Repetition(_) => true,
+                _ => false,
+            };
+
+            empty_run.push(e);
+            if !(fset.accepts_empty || is_rep) {
+                // this is the last one in the run
+                break;
+            }
+        }
+
+        if empty_run.len() > 0 {
+            self.0 = self.0 && !fsets_clash(&empty_run);
+        }
+    }
+
+    fn check_alternation(&mut self, es: &[Hir]) {
+        self.0 = self.0 && !fsets_clash(&es.iter().collect::<Vec<_>>());
```

Comment:
Not sure if it would be possible to change `fsets_clash` to always accept `&[Hir]` or that would require cloning in another callsite, but at least you can make it generic over both with something like:

```rust
fn fsets_clash<H: Borrow<Hir>>(es: &[H]) -> bool
```

Reply:
That sounds like exactly the sort of type magic I was trying to find. I just pushed a patch with my macro hack, but I'll give that a try soon.

Reply:
It looks like Borrow is more about smart pointers than what I want to do here.

Reply:
No, not necessarily. In my experience it's most common usecase is exactly accepting owned vs references to some type.